lib/auth/rotate.go

```diff
@@ -486,150 +486,99 @@ func (a *Server) startNewRotation(req rotationReq, ca types.CertAuthority) error
-	} else {
-		sshPrivateKey, sshCryptoSigner, err := a.keyStore.GenerateRSA()
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		sshSigner, err := ssh.NewSignerFromSigner(sshCryptoSigner)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		sshPublicKey := ssh.MarshalAuthorizedKey(sshSigner.PublicKey())
-		sshKeyPair = &types.SSHKeyPair{
-			PublicKey:      sshPublicKey,
-			PrivateKey:     sshPrivateKey,
-			PrivateKeyType: keystore.KeyType(sshPrivateKey),
+		if len(activeKeys.SSH) > 0 {
+			signer, err := ssh.NewSignerFromKey(rsaKey)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			sshPublicKey := ssh.MarshalAuthorizedKey(signer.PublicKey())
+			newKeySet.SSH = append(newKeySet.SSH, &types.SSHKeyPair{
+				PublicKey:      sshPublicKey,
+				PrivateKey:     req.privateKey,
+				PrivateKeyType: types.PrivateKeyType_RAW,
+			})
 		}
 
-		tlsPrivateKey, tlsSigner, err := a.keyStore.GenerateRSA()
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		tlsCert, err := tlsca.GenerateSelfSignedCAWithSigner(
-			tlsSigner,
-			pkix.Name{
-				CommonName:   ca.GetClusterName(),
-				Organization: []string{ca.GetClusterName()},
-			}, nil, defaults.CATTL)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		tlsKeyPair = &types.TLSKeyPair{
-			Cert:    tlsCert,
-			Key:     tlsPrivateKey,
-			KeyType: keystore.KeyType(tlsPrivateKey),
+		if len(activeKeys.TLS) > 0 {
+			tlsCert, err := tlsca.GenerateSelfSignedCAWithConfig(tlsca.GenerateCAConfig{
+				Signer: rsaKey.(*rsa.PrivateKey),
+				Entity: pkix.Name{
+					CommonName:   ca.GetClusterName(),
+					Organization: []string{ca.GetClusterName()},
+				},
+				TTL:   defaults.CATTL,
+				Clock: req.clock,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			newKeySet.TLS = append(newKeySet.TLS, &types.TLSKeyPair{
+				Cert:    tlsCert,
+				Key:     req.privateKey,
+				KeyType: types.PrivateKeyType_RAW,
+			})
 		}
 
-		jwtPrivateKey, jwtSigner, err := a.keyStore.GenerateRSA()
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		jwtPublicKey, err := utils.MarshalPublicKey(jwtSigner)
-		if err != nil {
-			return trace.Wrap(err)
+		if len(activeKeys.JWT) > 0 {
+			jwtPublicKey, jwtPrivateKey, err := utils.MarshalPrivateKey(rsaKey.(*rsa.PrivateKey))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			newKeySet.JWT = append(newKeySet.JWT, &types.JWTKeyPair{
+				PublicKey:      jwtPublicKey,
+				PrivateKey:     jwtPrivateKey,
+				PrivateKeyType: types.PrivateKeyType_RAW,
+			})
 		}
-		jwtKeyPair = &types.JWTKeyPair{
-			PublicKey:      jwtPublicKey,
-			PrivateKey:     jwtPrivateKey,
-			PrivateKeyType: keystore.KeyType(jwtPrivateKey),
+	} else {
+		if !additionalKeys.Empty() {
+			// Special case where a new HSM auth server is coming up and has
+			// already added local AdditionalTrustedKeys during the standby
+			// phase. Keep the existing AdditionalTrustedKeys to avoid
+			// invalidating the current Admin identity.
+			newKeySet = mergeKeySets(additionalKeys, newKeySet)
```

Comment:
`newKeySet` is empty here, right?
can it be 
```suggestion
			newKeySet = additionalKeys.Clone()
```